cmd/kubeadm/app/cmd/upgrade/diff.go

```diff
@@ -107,50 +114,49 @@ func validateManifestsPath(manifests ...string) (err error) {
 	return nil
 }
 
-func runDiff(flags *diffFlags, args []string) error {
-	var err error
-	var cfg *kubeadmapi.InitConfiguration
-	if flags.cfgPath != "" {
-		cfg, err = configutil.LoadInitConfigurationFromFile(flags.cfgPath, configutil.LoadOrDefaultConfigurationOptions{
-			SkipCRIDetect: true,
-		})
-	} else {
-		var client *client.Clientset
-		client, err = kubeconfigutil.ClientSetFromFile(flags.kubeConfigPath)
-		if err != nil {
-			return errors.Wrapf(err, "couldn't create a Kubernetes client from file %q", flags.kubeConfigPath)
-		}
-		cfg, err = configutil.FetchInitConfigurationFromCluster(client, nil, "upgrade/diff", false, false)
+// FetchInitConfigurationFunc defines the signature of the function which will fetch InitConfiguration from cluster.
+type FetchInitConfigurationFunc func(client clientset.Interface, printer output.Printer, logPrefix string, newControlPlane, skipComponentConfigs bool) (*kubeadmapi.InitConfiguration, error)
+
+func runDiff(fs *pflag.FlagSet, flags *diffFlags, args []string, fetchInitConfigurationFromCluster FetchInitConfigurationFunc) error {
+	upgradeCfg, err := configutil.LoadUpgradeConfig(flags.cfgPath)
+	if err != nil {
+		return err
+	}
+	client, err := kubeconfigutil.ClientSetFromFile(flags.kubeConfigPath)
+	if err != nil {
+		return errors.Wrapf(err, "couldn't create a Kubernetes client from file %q", flags.kubeConfigPath)
 	}
+	initCfg, err := fetchInitConfigurationFromCluster(client, &output.TextPrinter{}, "upgrade/diff", false, true)
 	if err != nil {
 		return err
 	}
 
-	// If the version is specified in config file, pick up that value.
-	if cfg.KubernetesVersion != "" {
-		flags.newK8sVersionStr = cfg.KubernetesVersion
+	// Pick up the version from the ClusterConfiguration.
+	if initCfg.KubernetesVersion != "" {
```

Comment:
for `kubeam diff` command, there are three priorities there: `ClusterConfiguration .kuberneteVersion` -> `upgradeCfg.Diff.KubernetesVersion` -> `version paramter`

Reply:
yes, param is the highest priority, then version in local config, then cluster config is lowest